### PR TITLE
PHPCS: nullable types PHP 8.4

### PR DESCRIPTION
**Pull request by andypost**
First part for #624

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -14,7 +14,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        php: ["7.2", "7.3", "7.4", "8.0", "8.1", "8.2", "8.3"]
+        php: ["7.2", "7.3", "7.4", "8.0", "8.1", "8.2", "8.3", "8.4"]
```
Comment (stof) [inline .github/workflows/build.yml:17]: This should not be done in this PR applying a CS change (especially as it does not work as is). Keep this PR focused only on the CS change.

**Review by stof — COMMENTED** [review]
This PR should also fix the new deprecations reported by a optional argument before required ones (which was skipped when the default value was making the type nullable implicitly).

Note that PHP-CS-Fixer has a (risky) rule `no_unreachable_default_argument_value` for that as well

File: .github/workflows/build.yml (hunk 1 above)
Comment (andypost) [reply .github/workflows/build.yml]: removed the change, supposed to try run pipelines

Comment (stof) [reply .github/workflows/build.yml]: I opened https://github.com/phpspec/prophecy/pull/631 to add new CI jobs

Comment (andypost) [reply .github/workflows/build.yml]: Rebased on top of it

**Comment by jrfnl** [comment]
> This PR should also fix the new deprecations reported by a optional argument before required ones (which was skipped when the default value was making the type nullable implicitly).

I scanned this branch with PHPCompatibility, which can check for both deprecations and this branch is clean of both of those issues.

**Comment by stof** [comment]
> Edit: at least it is at this time (when I scanned it).

It was not the case when I asked for that change, as shown by the static analysis CI job at that time: https://github.com/phpspec/prophecy/actions/runs/10492742258/job/29065879468

**Review by stof — APPROVED** [review]

**Pull request merged**